docs/proposals/block-storage-time-series-deletion.md

```diff
@@ -0,0 +1,272 @@
+
+The tombstone will be stored in a single file per request: 
+
+- `/<tenantId>/series_deletion/requests/<request_id>.json`
+
+
+The schema of the JSON file is:
+
+
+```
+{
+  "requestId": <string>,
+  "startTime": <int>,
+  "endTime": <int>,
+  "creationTime": <int>,
+  "matchers": [
+    "<string matcher 1>",
+    .., 
+    "<string matcher n>"
+    ]
+  },
+  "userID": <string>,
+  "deleteRequestStatus": <received | buildingPlan | deleting | processed> 
+}
+```
+
+
+Pros:
+
+- Design is similar to the existing chunk storage deletion 
+    - Lots of code can be reused inside the purger component. 
+- Allows deletion and un-delete to be done in a single operation. 
+
+Cons:
+- Negative impact on query performance when there are active tombstones. As in the chunk storage implementation, all the queries made will have to be compared to the matchers contained in the active tombstone files. The impact on performance should be the same as the deletion would have with chunk storage.
+
+
+#### Invalidating cache
+
+Using block store, the different caches available are:
+- Index cache
+- Metadata cache
+- Chunks cache (stores the potentially to be deleted chunks of data) 
+- Query results cache (stores the potentially to be deleted data) 
+
+The filtering using tombstones is only requested by the querier. By using the purger, the querier filters out the data received from the ingestors and store-gateway. The cache not being processed through the querier needs to be invalidated to prevent deleted data from coming up in queries. There are two potential caches that could contain deleted data, the chunks cache, and the query results cache. 
+
+
+Firstly, the query results cache needs to be invalidated for each new delete request. This can be done using the same mechanism currently used for chunk storage by utilizing the cache generation numbers. For each tenant, their cache is prefixed with a cache generation number. This is already implemented into the middleware and would be easy to use for invalidating the cache. When the cache needs to be invalidated due to a delete or cancel delete request, the cache generation numbers would be increased (to the current timestamp), which would invalidate all the cache entries for a given tenant. The cache generation numbers are currently being stored in an Index table  (e.g. DynamoDB or Bigtable). One option for block store is to store a per tenant JSON file with the corresponding cache generation numbers. 
```

Comment:
The concern with using KV-store is that it's not reliable. For instance consul is usually run as a single instance. If we loose the cache generation, won't the query-frontend return previously cached queries that might contain the deleted series?

Is storing the cache gen numbers in ingesters with replication an option?

Reply:
> The concern with using KV-store is that it's not reliable.

Let's say the cache gen number is the current timestamp. The query-frontend can check the KV-store and use the number from there. If not available (eg. consul was restarted from scratch and it's empty) the query-frontend can CAS a new gen number (which is the current timestamp). The side effect is that we'll have invalidated the query-frontend cache even if not required, but shouldn't cause other harms a part this one.

Reply:
> Is storing the cache gen numbers in ingesters with replication an option?

It would add ingesters dependency directly to query-frontend which I'm not sure it's something we want.

Reply:
Thank you for the suggestion, I've updated the proposal to include it.